### ISSUE TITLE
tls.createSecureContext needs the ca option for client cert authentication to work

### ISSUE THREAD

**Issue by mkremers** [Daplie/letsencrypt-express#35]
As of node version 4.3.1 the tls.createSecureContext function need the ca parameter set in order for client certificate authentication to work. In the cacheResult function of sni-callback.js the ca option is not present and should ideally trickle down from the ca set in httpsOptions.

**Comment by coolaj86** [comment]
*client certificate authentication* isn't something you generally use for a public-facing webserver (i.e. the kind you would use with Let's Encrypt) and the `ca` option has always been necessary to use it (as far as I can remember).

I can agree with you that perhaps `ca` should be copied over as a matter of completeness, however, are you sure that you actually intend to use *client certificate authentication* with Let's Encrypt? That's kinda like saying to your security guard: "Secure the building! Only let people in that have a t-shirt from hot topic!"... because anyone can just go get a t-shirt from hot topic, just like anyone can get a certificate from Let's Encrypt.

**Comment by mkremers** [comment]
Prior to node 4.3.1 you only needed to pass the ca to createServer for client authentication to work.

The use case i'm looking for is to grant specific rights to connections that present a client certificate signed by a corporate ca on a public facing website. The corporate ca needs to be passed to createSecureContext in order for the client certificate to be validated. A valid client cert is not required, it only grands specific access like "Only people with a Hot Topic t-shirt get access to the top floor" (and the t-shirt is not for sale, only given to employees of a corporation).

**Comment by coolaj86** [comment]
I can't work on this right now, but here's the path you'd need to follow to find and fix it yourself:

https://github.com/Daplie/letsencrypt-express/blob/master/lib/standalone.js#L171

https://github.com/Daplie/letsencrypt-express/blob/master/lib/standalone.js#L255

https://github.com/Daplie/letsencrypt-express/blob/master/lib/standalone.js#L108

https://github.com/Daplie/letsencrypt-express/blob/master/lib/standalone.js#L232

**Comment by coolaj86** [comment]
And actually, if you just skip the sugar and write a few more lines of code you can add the ca yourself:

https://github.com/Daplie/letsencrypt-express#use-with-raw-http--https-modules

Let me know if that's the solution you need or not.

**Comment by mkremers** [comment]
I added a pull request (#37) containing a one line fix to solve this issue.
This will basically append the ca's for the createSecureContext call when a ca was set on the httpsOptions object (requestCert would also need to be set to true anyway). I hope that this is an acceptable solution for inclusion in your project.

**Comment by coolaj86** [comment]
Thanks for the PR.

Are there any options that should be passed before I publish a new version?

**Comment by mkremers** [comment]
No options needed, its a pure "internal" change.

Thanks for including the patch for a new version!

**Comment by coolaj86** [comment]
Hey, I just reread some of your comments and realized that I interpreted incorrectly the first time. I didn't read that it was the `SNICallback` that was the problem and was just thinking `httpsOptions` to `createServer`.

Anyway, now that I was looking at it more carefully I see that what you were saying made sense the whole time and my comments back to you did not.  :$

Anyway, published as `v1.1.5` and I also copied over `crl`, `requestCert`, and `rejectUnauthorized`, just in case.

My spidey-sense is tingling telling me that the default value for `ca` should be either `undefined` or `require('ssl-root-cas').rootCas`, but not `certInfo.ca`.

Would you mind refreshing me on why `certInfo.ca` is the default value?

**Comment by mkremers** [comment]
It's easy to misinterpret someone, especially on a complex topic like encryption. I'm glad that it all worked out :-)

Regarding the default value for  `ca`, now that you made me think about it i'm not sure that the default value of `certInfo.ca` is correct. The reason I implemented it like this was because during debugging, the first moment that the `certInfo` object is passed into the module, the `ca` attribute is set to a value that I did not recognized as something from my system (a single ca cert and not a chain). I assumed that its a lets-encrypt ca but i did not check. It seemed to me the safe option is to pass the value as is or append it with `httpsOptions.ca`.

I'm not sure that adding `require('ssl-root-cas').rootCas` only for the `SNICallback` callback is a good idea. One can always do that as an option to `createServer` if needed. But that might be different for the other use-cases that you support with letsencrypt-express.
